src/g_kbdnav.h

```diff
@@ -0,0 +1,165 @@
+#ifndef HAVE_KEYBOARDNAV
+        #define HAVE_KEYBOARDNAV 1
+#endif
+
+#ifndef G_KBDNAV_H
+#define G_KBDNAV_H
+
+typedef struct _kbdnav
+{
+    int kn_state;        /* see KN_ consts below */
+    int kn_ioindex;      /* selected inlet/outlet, 0-based*/
+    int kn_moddown;      /* is our modifier key down? */
+    int kn_connindex;    /* selected in/out connection, 0-based */
+    t_outconnect *kn_outconnect;
+    t_object *kn_selobj; /* currently selected object */
+    char* kn_highlight;  /* highlight_color */
+    struct _linetraverser *kn_linetraverser;
+    struct _magtrav *kn_magtrav; /* linked list of connections for the magnetic connector*/
+    int kn_waitingfornumber;
+    int kn_chosennumber; /* used by the digit connector */
+    int kn_iotype;       /* IO_INLET/IO_OUTLET */
+    int kn_indexvis;     /* boolean: should we visualize object indexes? */
+} t_kbdnav;
+
+/* #define constants */
+
+#define KN_INACTIVE 0 /* not navigating with the keyboard */
+#define KN_IO_SELECTED 1 /* selecting an inlet/outlet */
+#define KN_CONN_SELECTED 2 /* navigating through an in/outlet's connections */
+#define KN_WAITING_NUMBER 3 /* waiting for a number from 0-9 when using the "digit connector" */
+
+#define KBDNAV_SEL_PAD_X 4
+#define KBDNAV_SEL_PAD_Y 3
+
+#define IO_INLET 0
+#define IO_OUTLET 1
+
+/* FORWARD DECLARATIONS */
+
+/* stuff from g_editor.c */
+
+EXTERN void canvas_connect_with_undo(t_canvas *x, t_float index1, t_float outno,
+                                         t_float index2, t_float inno);
+EXTERN void canvas_disconnect_with_undo(t_canvas *x, t_float index1, t_float outno,
+                                         t_float index2, t_float inno);
+EXTERN void canvas_reselect(t_canvas *x);
+EXTERN void canvas_clearline(t_canvas *x);
+
+
+/* g_editor.c */
```

Comment:
these generally useful functions should probably just go into `g_canvas.h` (and be marked as `EXTERN`).

(the `kbdnav_*`-functions should stay here and be kept private)

Reply:
ok 8c4bcbed8c7400d3c213b87c00b611e6e770c49b